create proper server socket error

Raynos opened the pull request:
TChannel used to eat server errors and much sadness would be
had when you tried to figure out why your tests do not work.

My emitting an error we have insight into what's going on.

```
AUTOBAHN 30809: error: null server socket error: { [Error: tchannel: listen EADDRINUSE]
  code: 'EADDRINUSE',
  errno: 'EADDRINUSE',
  syscall: 'listen',
  requestedPort: 36431,
  host: '10.32.163.144' } ~ null

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: tchannel: listen EADDRINUSE
    at Server.onServerSocketError (/home/raynos/uber/autobahn/node_modules/tchannel/index.js:98:27)
    at Server.EventEmitter.emit (events.js:95:17)
    at net.js:1046:12
    at process._tickDomainCallback (node.js:459:13)
```

Above is an example of the print statement of this branch.

Here we see what the failure is and because we create a new error we see a useful stack trace. Also by prefixing the EADDRINUSE with `tchannel` we know its a `listen()` failure from a tchannel server.

reviewers: @jcorbin @kriskowal

kriskowal commented on node/index.js:
```diff
@@ -105,7 +105,15 @@ function TChannel(options) {
+        serverError.requestedPort = self.requestedPort;
+        serverError.host = self.host;
+
+        self.logger.error(self.hostPort + ' server socket error: ' + inspect(serverError));
```
Dynamic string. Pass serverError properties as second arg.

Raynos commented:
@kriskowal @jcorbin Updated this PR.

I fixed the `logger.error()` callsite and added a `WrappedError` base class to dry up wrapping node core IO errors in a typederror thingy.

Raynos commented on node/index.js:
```diff
@@ -105,7 +112,20 @@ function TChannel(options) {
     });
     self.serverSocket.on('error', function onServerSocketError(err) {
-        self.logger.error(self.hostPort + ' server socket error: ' + inspect(err));
+        if (err.code === 'EADDRINUSE') {
```
note: we should only wrap EADDRINUSE errors; any other errors will be bubbled up without wrapping context.

Maybe we should wrap them in a generic `TChannelServerError` instead ...

jcorbin commented on node/test/regression-listening-on-used-port.js:
```diff
@@ -0,0 +1,34 @@
+    otherServer.listen(0, 'localhost');
+
+    function onPortAllocated() {
+        server.on('error', onError);
```
once?

Raynos replied:
`once('error')` causes uncaught exceptions. it's bad pattern

jcorbin commented on node/test/regression-listening-on-used-port.js:
```diff
@@ -0,0 +1,34 @@
+        assert.notEqual(-1, err.origMessage
+            .indexOf('listen EADDRINUSE'));
+
+        otherServer.close();
```
does close take a callback? can it be `server.close(assert.end)` ?

Raynos replied:
sure. Doesn't really matter though.

jcorbin commented:
:+1: , feedback to taste
